code/datums/components/barber.dm

```diff
@@ -448,6 +420,154 @@ TYPEINFO(/datum/component/barber)
+		src.preview.mirror_appearance(barbee)
+		src.preview.update_appearance(src.barbee.bioHolder.mobAppearance, direction=SOUTH, name=src.barbee.name)
+
+	var/list/current_hair_style = list("bottom" = new_AH.customization_first.name, "middle" = new_AH.customization_second.name, "top" = new_AH.customization_third.name)
+	. = list("preview" = src.preview.preview_id, "selected_hair_portion" = hair_portion, "current_hair_style" = current_hair_style)
+
+/datum/component/barber/ui_static_data(mob/user)
+	var/list/all_hairs = list()
+	var/all_styles = null
+
+	if (istype(src, /datum/component/barber/shave))
+		all_styles = concrete_typesof(/datum/customization_style/beard) + concrete_typesof(/datum/customization_style/moustache) + concrete_typesof(/datum/customization_style/sideburns) + concrete_typesof(/datum/customization_style/eyebrows)
+	else
+		all_styles = concrete_typesof(/datum/customization_style/hair)
+
+	// just so we get a special icon sprite for no hair
+	all_hairs += list("None" = list("hair_id" = "none", "hair_icon" = "data:image/png;base64," + icon2base64(icon('icons/map-editing/landmarks.dmi', "x", SOUTH))))
+
+	for (var/datum/customization_style/styles as anything in all_styles)
+		var/datum/customization_style/style = new styles
+		var/hair_icon = "data:image/png;base64," + icon2base64(icon('icons/mob/human_hair.dmi', style.id, SOUTH, 1)) // yeah, sure, i'll keep it white. the user can preview the hair style anyway.
+		all_hairs += list(style.name = list("hair_id" = style.id, "hair_icon" = hair_icon))
+
+	. = list("available_styles" = all_hairs)
+
+/datum/component/barber/ui_act(var/action, var/params)
+	. = ..()
+	if (.)
+		return
+
+	switch(action)
+
+		if("change_hair_portion")
+			src.hair_portion = params["new_portion"]
+			return TRUE
+
+		if("do_hair")
+			if (isnull(params["style_id"])) // It means we are making a wig
+				if (istype(src, /datum/component/barber/shave))
+					actions.start_and_wait(new/datum/action/bar/barber/shave(src.barbee, src.barber, get_barbery_conditions(src.barbee, src.barber), null, ALL_HAIR), src.barber)
+				else
+					actions.start_and_wait(new/datum/action/bar/barber/haircut(src.barbee, src.barber, get_barbery_conditions(src.barbee, src.barber), null, ALL_HAIR), src.barber)
+
+				src.new_AH.CopyOther(src.barbee.bioHolder.mobAppearance)
+				src.preview.mirror_appearance(barbee)
+				src.preview.update_appearance(new_AH)
+				return TRUE
+
+			var/hair_portion_list = list(
+				"bottom" = BOTTOM_DETAIL,
+				"middle" = MIDDLE_DETAIL,
+				"top" = TOP_DETAIL
+			)
+
+			var/hair_portion_selected = hair_portion_list[src.hair_portion]
+			var/datum/customization_style/new_hairstyle = null
+			var/all_hairs = null
+
+			if (istype(src, /datum/component/barber/shave))
+				all_hairs = concrete_typesof(/datum/customization_style/beard) + concrete_typesof(/datum/customization_style/moustache) + concrete_typesof(/datum/customization_style/sideburns) + concrete_typesof(/datum/customization_style/eyebrows)
```

Comment:
Rather than make this every time, and in `ui_static_data`, would it make sense to create when the datum is initialized and refer to here?